src/lib/transaction.ts

```diff
@@ -16,40 +16,66 @@ import {
+import { deriveKey, getKey } from './key'
 import { listUnspents } from './wallet'
 import { TransactionBuilder } from 'bitcoinjs-lib';
-import { ListUnspentsBackendResponse } from 'response';
+import { ListUnspentsBackendResponse, GetWalletBackendResponse, GetKeyBackendResponse } from 'response';
 import BigNumber from "bignumber.js";
 import { btcToSatoshi } from './utils/helpers'
+import { decrypt } from './crypto';
+import { ErrorResponse } from '../types/response'
 
 const joinPath = (path: Path): string =>
   `${path.cosignerIndex}/${path.change}/${path.addressIndex}`
 
-export const sendCoins = async (
-  userToken: string, xprv: string, walletId: string, recipients: Recipient[]
+export const send = async (
+  userToken: string, walletId: string, recipients: Recipient[], xprv?: string, password?: string
 ): Promise<void> => {
-  const txb = initializeTxBuilder()
   const { recommended } = await getFeesRates()
   const unspentsResponse = await listUnspents(userToken, walletId, recommended, recipients)
   const wallet = await getWallet(userToken, walletId)
   const pubKeys = wallet.keys.map((key: Key) => key.pubKey)
   const changeAddresResponse = await createNewAddress(userToken, walletId, true)
+  const userXprv = await xprivOrGetFromServer(userToken, wallet, xprv, password)
+  const txHex = buildTxHex(unspentsResponse, recipients, userXprv, changeAddresResponse.address, pubKeys)
+  await sendTransaction(userToken, walletId, txHex)
+}
 
+const buildTxHex = (unspentsResponse: ListUnspentsBackendResponse, recipients: Recipient[], xprv: string, changeAddres: string, pubKeys: string[]): string => {
+  const txb = initializeTxBuilder()
   const inputs = sortUnspents(unspentsResponse.outputs)
   inputs.forEach((uns: UTXO) => {
     txb.addInput(uns.txHash, uns.n)
   })
 
-  const outputs = createOutputs(unspentsResponse, recipients, changeAddresResponse.address)
+  const outputs = createOutputs(unspentsResponse, recipients, changeAddres)
   outputs.forEach((out: TxOut) => {
     txb.addOutput(out.script, out.value.toNumber())
   })
 
   signInputs(inputs, xprv, pubKeys, txb)
+  return txb.build().toHex()
+}
 
-  const txHex = txb.build().toHex()
-  await sendTransaction(userToken, walletId, txHex)
+const xprivOrGetFromServer = async (userToken: string, wallet: GetWalletBackendResponse, xprv?: string, password?: string): Promise<string> => {
+  if (xprv) {
+    return xprv
+  } else if (password) {
+    return await getUserXprvFromServer(wallet, userToken, password)
+  } else {
+    throw <ErrorResponse>({ message: "Password or xprv has to be specified!", code: 400 })
+  }
+}
+
+const getUserXprvFromServer = async (wallet: GetWalletBackendResponse, userToken: string, password: string): Promise<string> => {
+  const keyId: string = wallet.keys.find(key => key.type === KeyType.USER)!.id
+  const key: GetKeyBackendResponse = await getKey(userToken, keyId, true)
+  const prvKey = key.prvKey
+  if (prvKey) {
+    return decrypt(password, prvKey)
```

Comment:
if there's a decryption error, will the user get a nice error message?

Reply:
I've just fixed it ;)